### ISSUE TITLE
rule-selector-property-value-disallowed-list

### ISSUE THREAD

**Issue by clshortfuse** [stylelint/stylelint#7054]
### What is the problem you're trying to solve?

I would like to disallow

* a specific value
* on a specific property
* in a specific rule 

Specifically, I would like to disallow `overflow` everywhere except for `:host` while still allowing `overflow:hidden` everywhere.

### What solution would you like to see?

This is a combination of the logic of `rule-selector-property-disallowed-list` and `declaration-property-value-disallowed-list`

Basically, something like 

````json
"rule-selector-property-value-disallowed-list": [
  { "/^(?!:host)/": ["/^overflow(-[xy])?$/", "/^(?!hidden).+/"] }
],
````

**Comment by jeddy3** [comment]
@clshortfuse Thanks for the request and for using the template.

I think we can add a rule to do this, although it may be more suitable as a [plugin](https://stylelint.io/developer-guide/plugins) as it [may not be generally useful](https://stylelint.io/developer-guide/rules#define-the-rule).

I'll label as discussion in case anyone else wants to chime in.

If we did add it, we might want to tweak the options structure so that the `value-disallowed-list` is consistent across rules:

```
"rule-selector-property-value-disallowed-list": [
  { "/^(?!:host)/": [ { "/^overflow(-[xy])?$/": ["/^(?!hidden).+/"] } ]
],
```

i.e. the value of each object in the array is an array of objects

**Comment by jeddy3** [comment]
- Name: `rule-selector-property-value-disallowed-list`
- Primary option: `object { "selector": [{ "property"|/regex/|"/regex/": "value"|"/regex/"|/regex/) }] }`
- Secondary options: None
- Autofixable: No
- Message: `Unexpected value "${value}" for property "${property}" and selector "${selector}"`
- Description: "Specify a list of disallowed values for properties and selectors within rules."
- Section: "Enforce conventions" -> "Allowed, disallowed & required" -> "Rule"
- Config: No

I've labelled the issue as ready to implement. Please consider [contributing](https://stylelint.io/contributing) if you have time.

There are [steps on how to add a new rule](https://stylelint.io/developer-guide/rules#add-a-rule) in the Developer guide.

You can use [`rule-selector-property-disallowed-list`](https://github.com/stylelint/stylelint/tree/main/lib/rules/rule-selector-property-disallowed-list) as a blueprint.

**Comment by ybiquitous** [comment]
@jeddy3 Thanks for summarizing. But I'd like to confirm the rule guideline before implementing a new rule. The guideline says "*A rule must have singular purpose that doesn't overlap with other rules*":

https://github.com/stylelint/stylelint/blob/d03def645ef8c67446ccb86285566f811372b9ca/docs/developer-guide/rules.md?plain=1#L19

And please see the config example below, using `rule-selector-property-value-disallowed-list` and `rule-selector-property-disallowed-list`. Are these rules overlapped? And will users not be confused?

```js
{
  // New rule
  'rule-selector-property-value-disallowed-list': {
    a: [{ color: /.+/ }]
  },

  // Existing rule: the same as above?
  'rule-selector-property-disallowed-list': {
    a: ['color']
  }
}
```

Otherwise, maybe we might be able to extend the primary option of `rule-selector-property-disallowed-list`, e.g.

```js
{
  // Extend the primary option
  'rule-selector-property-disallowed-list': {
    a: ['color'],          // 'a' selector must not have 'color' property (current)
    b: [{ color: 'red' }]  // 'b' selector must not have 'color' property with 'red' value (new)
  }
}
```

What do you think?

**Comment by Mouvedia** [comment]
> …we might be able to extend the primary option of `rule-selector-property-disallowed-list`…
> `b: [{ color: 'red' }]`

If the extension of the option gets added, it should detect useless properties.
e.g.
```
    a: ['color', { color: 'red' }],
```
should warn that `color: 'red'` is unnecessary because `'color'` is already not allowed.
<hr/>
Do we allow multiple objects?
e.g.

```
// if we have
a: [{ position: 'static' }, { color: 'red' }],
// do we error out "please merge the objects"?
// i.e. it should be
a: [{ position: 'static', color: 'red' }],
```

**Comment by ybiquitous** [comment]
@Mouvedia Good point. So how about the following idea?

```js
{
  'a': 'color',
  'b': /color/,
  'c': ['color', /position/],
  'd': {'color': 'red', 'color': /blue/},
}

/* ⬇️ noramilzed ⬇️ */

{
  'a': {'color': /.*/},
  'b': {'/color/': /.*/},
  'c': {'color': /.*/, '/position/': /.*/},
  'd': {'color': /blue/},  // normal JS object's behavior
}
```

**Comment by Mouvedia** [comment]
> So how about the following idea?

That's much better and doesn't require to document the merge/overwrite strategy.
i.e. having duplicate properties is a syntax error in strict mode
